### PR TITLE
Handle dangling symlinks in outputs for incremental build

### LINKED ISSUE #9708: Handle dangling symlinks in outputs for incremental build
When snapshotting we sometimes fail if we encounter a dangling symlink.
We want to change this, and treat a dangling symlink as if we encountered a missing file.
In an incremental build, if the dangling symlink becomes a non-dangling one (or vice versa), we consider that a change similar to changing a file into a directory.

* [ ] task produces output (as output file/in output directory) with a dangling symlink
  * [x] should succeed (fails currently)
  * [x] should be up-to-date without changes
  * [x] changing the output to a non-dangling symlink makes the task output out of date (should track Missing files in Snapshots/Fingerprints)

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes #9708

### CODE REVIEW

File: subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputFilterUtil.java [hunk 1]
```diff
@@ -176,6 +182,11 @@ public void postVisitDirectory(DirectorySnapshot directorySnapshot) {
                 currentRoot = null;
             }
         }
+
+
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputFilterUtil.java]: Nit: additional new line.

File: subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java [hunk 2]
```diff
@@ -67,7 +69,7 @@ public boolean preVisitDirectory(DirectorySnapshot directorySnapshot) {
 
                 @Override
                 public void visitFile(FileSystemLocationSnapshot fileSnapshot) {
-                    if (!includeMissing && fileSnapshot.getType() == FileType.Missing) {
+                    if (isRoot() && (!includeMissing && fileSnapshot.getType() == FileType.Missing)) {
```
Comment (wolfs) [inline subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java]:
```suggestion
                    if (isRoot() && !includeMissing && fileSnapshot.getType() == FileType.Missing) {
```

I think the brackets are unnecessary.

File: subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy [hunk 3]
```diff
@@ -72,11 +72,26 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
...
-    def "detect changes to broken symlink outputs in #outputType"() {
+    def "detect changes to broken symlink outputs in OutputDirectory"() {
         def root = file("root").createDir()
         def target = file("target")
         def link = root.file("link")
-        buildFile << script(target, link)
+        buildFile << """
+            import java.nio.file.*
+            class ProducesLink extends DefaultTask {
+                @OutputDirectory File outputDirectory
```
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy]: Any reason you use a `File` here and `Path` in `detect changes to broken symlink outputs in OutputFile`?

File: subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java (hunk 2 above)
Comment (paplorinc) [reply subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java]: Just wanted to group the include missing flag with checking for missing, to make sure they're interpreted together. Would you like me to split it?

Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java]: I think `includeMissing` means now `includeMissingRoots` and I would probably move it to the first position. I don't think the brackets help much in interpreting the boolean condition.

Comment (paplorinc) [reply subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java]: If you don't think it helps, I'll change it :)

Comment (paplorinc) [reply subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java]: Done

File: subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy (hunk 3 above)
Comment (paplorinc) [reply subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy]: I'm not referencing it directly here, and need it to be Path in the other. If diversity is not a virtue here, I'll change both to `Path`.

Comment (paplorinc) [reply subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy]: Done

File: subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputFilterUtil.java (hunk 1 above)
Comment (paplorinc) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputFilterUtil.java]: Done

File: subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java (hunk 2 above)
Comment (paplorinc) [reply subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java]: Renamed to `includeMissingRoots`

**Comment by paplorinc** [comment]
Merged, thanks for the guidance.

**Pull request merged**